src/backend/optimizer/plan/planner.c

```diff
@@ -908,6 +908,14 @@ subquery_planner(PlannerGlobal *glob, Query *parse,
 											 list_make1_int(root->is_split_update),
 											 rowMarks,
 											 SS_assign_special_param(root));
+			/*
+			 * Set up the visible plan targetlist as being the same as
+			 * the RETURNING list. This is for the use of
+			 * EXPLAIN; the executor won't pay any attention to the
+			 * targetlist. Also the targetlist might be needed by higher-level nodes,
+			 * such as Material, for correct operation.
+			 */
+			plan->targetlist = copyObject(parse->returningList);
```

Comment:
The make_modifytable function has a returningLists argument, which is passed parse->returningList. Why not do both changes in the same place in the make_modifytable function?

Reply:
make_modifytable takes a list of lists, and it's not clear which one should be used.

Reply:
Solution is partial revert https://github.com/arenadata/gpdb/commit/9fa82c980935ef4aee18fabe8da20ae2198b052a but why?!
```diff
diff --git a/src/backend/optimizer/plan/createplan.c b/src/backend/optimizer/plan/createplan.c
index 986b27889b..5b453ca2c0 100644
--- a/src/backend/optimizer/plan/createplan.c
+++ b/src/backend/optimizer/plan/createplan.c
@@ -6367,8 +6367,10 @@ make_modifytable(PlannerInfo *root,
 	node->plan.lefttree = NULL;
 	node->plan.righttree = NULL;
 	node->plan.qual = NIL;
-	/* setrefs.c will fill in the targetlist, if needed */
-	node->plan.targetlist = NIL;
+	if (returningLists)
+		node->plan.targetlist = copyObject(linitial(returningLists));
+	else
+		node->plan.targetlist = NIL;
 
 	node->operation = operation;
 	node->canSetTag = canSetTag;
diff --git a/src/backend/optimizer/plan/setrefs.c b/src/backend/optimizer/plan/setrefs.c
index 98eefd64e8..39f7f95208 100644
--- a/src/backend/optimizer/plan/setrefs.c
+++ b/src/backend/optimizer/plan/setrefs.c
@@ -1044,7 +1044,6 @@ set_plan_refs(PlannerInfo *root, Plan *plan, int rtoffset)
 			{
 				ModifyTable *splan = (ModifyTable *) plan;
 
-				Assert(splan->plan.targetlist == NIL);
 				Assert(splan->plan.qual == NIL);
 
 				splan->withCheckOptionLists =
@@ -1088,7 +1087,6 @@ set_plan_refs(PlannerInfo *root, Plan *plan, int rtoffset)
 					 * we don't have to do set_returning_clause_references()
 					 * twice on identical targetlists.
 					 */
-					splan->plan.targetlist = copyObject(linitial(newRL));
 				}
 
 				foreach(l, splan->resultRelations)
```